common/src/main/java/org/fiware/tmforum/common/rest/AbstractSubscriptionApiController.java

```diff
@@ -0,0 +1,67 @@
+import org.fiware.tmforum.common.repository.TmForumRepository;
+import org.fiware.tmforum.common.util.StringUtils;
+import org.fiware.tmforum.common.validation.ReferenceValidationService;
+import reactor.core.publisher.Mono;
+
+import java.net.URI;
+import java.util.List;
+import java.util.Map;
+import java.util.UUID;
+
+import static org.fiware.tmforum.common.CommonConstants.DEFAULT_OFFSET;
+
+@Slf4j
+public abstract class AbstractSubscriptionApiController extends AbstractApiController<Subscription> {
+    private final Map<String, String> eventGroupToEntityNameMapping;
+
+    public AbstractSubscriptionApiController(ReferenceValidationService validationService, TmForumRepository repository,
+                                             Map<String, String> eventGroupToEntityNameMapping) {
+        super(validationService, repository);
+        this.eventGroupToEntityNameMapping = eventGroupToEntityNameMapping;
+    }
+
+    protected Mono<Subscription> create(Subscription subscription) {
+        return findExistingSubscription(subscription)
+                .switchIfEmpty(create(Mono.just(subscription), Subscription.class));
+    }
+
+    private Mono<Subscription> findExistingSubscription(Subscription subscription) {
+        String query = String.format("callback==\"%s\";query==\"%s\";%s",
+                subscription.getCallback(), subscription.getQuery(),
+                String.join(";", subscription.getEventTypes().stream().map(eventType ->
+                    String.format("eventTypes==\"%s\"", eventType)).toList()));
+        return repository.findEntities(DEFAULT_OFFSET, 1, Subscription.TYPE_SUBSCRIPTION,
+                    Subscription.class, query)
+                .flatMap(list -> list.isEmpty() ? Mono.empty() : Mono.just(list.get(0)));
+    }
+
+    protected Subscription buildSubscription(String callback, String query, List<String> eventGroups) {
+        NotificationQuery notificationQuery = QueryParser.parseNotificationQuery(query, eventGroups);
+        log.debug(notificationQuery.toString());
+
+        List<String> eventTypes;
+        if (notificationQuery.getEventTypes().isEmpty()) {
+            eventTypes = EventConstants.ALLOWED_EVENT_TYPES.get(notificationQuery.getEventGroupName()).stream().map(
+                    eventType -> notificationQuery.getEventGroupName() + eventType).toList();
+        } else {
+            eventTypes = notificationQuery.getEventTypes();
+        }
+
+        String subId = UUID.randomUUID().toString();
+        Subscription subscription = new Subscription(subId);
+        subscription.setEventTypes(eventTypes);
+        subscription.setEntities(List.of(eventGroupToEntityNameMapping.get(notificationQuery.getEventGroupName())));
+        subscription.setQuery(notificationQuery.getQuery());
+        subscription.setPayloadName(StringUtils.decapitalize(notificationQuery.getEventGroupName()));
+        subscription.setCallback(URI.create(callback));
```

Comment:
I thought we decided that the TM Forum API Service receives the notifications from ngsi-ld and maps them to the format defined by TM Forum? It would be nice if the broker could send them out directly but I thought we decided against it, since we can't hand over the template properly?